streamz/tests/test_kafka.py

```diff
@@ -553,3 +553,75 @@ def test_kafka_batch_checkpointing_async_nodes_2():
         assert committed2[1].offset == 2
         assert committed3[0].offset == 1
         assert committed3[1].offset == 1
+
+
+def test_kafka_checkpointing_auto_offset_reset_latest():
+    '''
+    Testing whether checkpointing works as expected with multiple topic partitions and
+    with auto.offset.reset configuration set to latest (also default).
+    '''
+    j = random.randint(0, 10000)
+    ARGS = {'bootstrap.servers': 'localhost:9092',
+            'group.id': 'streamz-test%i' % j,
+            'auto.offset.reset': 'latest'}
+    with kafka_service() as kafka:
+        kafka, TOPIC = kafka
+        TOPIC = "test-checkpointing-offset-reset-latest"
+        subprocess.call(shlex.split("docker exec streamz-kafka "
+                                    "/opt/kafka_2.11-0.10.1.0/bin/kafka-topics.sh "
+                                    "--create --zookeeper localhost:2181 "
+                                    "--replication-factor 1 --partitions 3 "
+                                    "--topic test-checkpointing-offset-reset-latest"))
+        time.sleep(2)
+
+        '''
+        Since the stream has not started yet, these messages are not read because
+        the stream has auto.offset.reset set to latest.
+        '''
+        for i in range(30):
+            kafka.produce(TOPIC, b'value-%d' % i)
+        kafka.flush()
+
+        stream1 = Stream.from_kafka_batched(TOPIC, ARGS, asynchronous=True)
+        out1 = stream1.map(split).gather().sink_to_list()
+        stream1.start()
+        time.sleep(5)
```

Comment:
It should not be necessary to put all these `sleep` statements in here. If necessary, we could have wait_for loops that poll for successful starting.

Reply:
What predicate do you suggest using in the `wait_for` in this case?

Reply:
Can't think of anything else here other than sleep to ensure that the messages aren't produced to Kafka before the stream successfully starts.

Reply:
What do you think? Is it okay to leave this as is? I see a couple of other tests using sleep too.

Reply:
Is there a way to poll with the producer to see if the broker is ready to accept messages?

Reply:
Or the consumer, I suppose: as soon as list_topics doesn't fail, and includes the topic we're after, we should be ready to start. This kind of logic can live in the kafka fixture too, or a utility function.

Note that the stream's `start` method is blocking, we shouldn't need to wait. It adds the poll_kafka callback, so we could asyncio.sleep(0) to make sure that runs - evrything in it is blocking until the poll is complete, except for possible checkpointing, which will have nothing to do at this point.

Reply:
> Is there a way to poll with the producer to see if the broker is ready to accept messages?

Not that I know of. I'm assuming you're referring to sleep on line 575. Typically, the creation of a topic is blocking, so I'll remove this sleep. 

> Note that the stream's start method is blocking, we shouldn't need to wait. It adds the poll_kafka callback, so we could asyncio.sleep(0) to make sure that runs - evrything in it is blocking until the poll is complete, except for possible checkpointing, which will have nothing to do at this point.

The point is to ensure that the [first polling cycle](https://github.com/python-streamz/streamz/blob/7f92174ae4c21a60105e1b8476ae3936c5f1c6d3/streamz/sources.py#L514), which includes the execution of lines 530 and the branch on line 534, should finish before we produce any new messages to Kafka. Removing sleep is causing the producer to produce before the stream can finish the first poll cycle. Note that this is just so that we can simulate a good scenario to test the functionality where auto.offset.reset is latest. Let me know if this makes sense.

Reply:
so a single asyncio/gen-sleep of zero time should do, as start already queues the polling coroutine, and the first call of it is guaranteed to get that far.

Reply:
Nope, using `gen.sleep(0 or 5)` doesn't seem to work, the producer still produces before the first cycle. I think we need a blocking call to ensure the stream completes the first poll cycle.

Reply:
This maybe due to the `list_topics`, `get_watermark_offsets` being relatively expensive calls.

Reply:
It would be really good to figure out exactly why, and this would go a great distance to fixing the various flaky/unstable kafka pieces. I know it's complicated: streamz's event loop schedules when the coroutine runs, but rdkafka is actually in a different (non-python) thread.

Perhaps the source class could have a `running` instance attribute which is initially `False`, but it gets set the first time that `get_watermark_offsets` succeeds (i.e., after the `except` block) - or let `stopped` have more than `True/False` values.

Reply:
^ this, so that we have something whose status we can (a)wait on.

Reply:
> this would go a great distance to fixing the various flaky/unstable kafka pieces

I agree.

> Perhaps the source class could have a running instance attribute which is initially False, but it gets set the first time that get_watermark_offsets succeeds (i.e., after the except block) - or let stopped have more than True/False values.

This is definitely a way out, although I'm not sure if we ever need it except for simulating tests with `auto.offset.reset:latest` Kafka config. I could simplify the tests a little, but I was thinking the tests should cover the config exhaustively (at the risk of some sleep statements, for now), especially after I introduced this nasty bug. 

But I agree, it would definitely be worth spending some time figuring our the flaky nature of some of the tests.

Reply:
The trouble is, as we have found, that `sleep` doesn't performs the same on CI as locally, doubtless because of the puny workers that the tests run on. The totality of a `sleep` may sometimes be eaten up by garbage collection or perhaps some other build on the same host. I addition, having a test that takes a minimum of tens of seconds to run is unfortunate.

I do agree with the desire to test all configs.

Reply:
Tests should always be predictable so you don't get intermittent failures. You should not use `random` unless it is used to pre-generate test data and using `sleep` is more or less just a patch. The best solution is to figure out how to watch or listen to the event you're waiting for while sleeping, like what the `wait_for` function is used for.

Reply:
@martindurant, would we need to make some changes to the `client.submit` calls in streamz to make sure it works with the Dask [changes](https://github.com/dask/distributed/issues/4177)?

Reply:
Indeed maybe so, but this should be a follow-up. It seems we may need to convert some tuples to lists, but we can wait until things start to fail, I think.

Reply:
I think CI was already failing in the last couple of builds:

https://travis-ci.org/github/python-streamz/streamz/builds/737860381
https://travis-ci.org/github/python-streamz/streamz/builds/738116645

Reply:
I believe the immediate change will be reverted, but let's see what decision is made on that PR. If you have an opinion, can voice it there.

Reply:
Ah, I see that temporary fix just got merged. Hopefully we can trigger a build after some time to see if CI passes here?

Reply:
Maybe the fix isn't in yet? CI is failing with this error: `ImportError: cannot import name 'Layer' from 'dask.highlevelgraph'`.

Reply:
@martindurant, any ideas?